// @corva/create-app/lib/scripts/utils/version.js
import NpmApi from 'npm-api';
import chalk from 'chalk';
import semver from 'semver';
import inquirer from 'inquirer';

import packageJson from '../../../package.json' assert { type: 'json' };
import { logger } from '../../helpers/logger.js';

const npm = new NpmApi();

const error = chalk.bold.red;
const warning = chalk.hex('#FFA500'); // Orange
const code = chalk.cyan;
const asterisks = '****************************************************************';

const getCurrentVersion = () => packageJson.version;
const getLatestVersion = async () => npm.repo('@corva/create-app').prop('version');

// NOTE: Stop process and show error if version is outdated
export async function ensureLatestVersion() {
  const currentVersion = getCurrentVersion();
  const latestVersion = await getLatestVersion();

  const isCurrentVersionOutdated = semver.gt(latestVersion, currentVersion);

  if (isCurrentVersionOutdated) {
    console.log(error(asterisks));
    console.log(error`
      Your version of the @corva/create-app is outdated.
      There is a new version available ({green.bold ${latestVersion}}).
      Please update the @corva/create-app with the following command:

          ${code('npm i -g @corva/create-app@latest')}

    `);
    console.log(error(asterisks));
    process.exit(0);
  }
}

// NOTE: Show user-friendly warning if version is outdated
export async function warnIfOutdated() {
  logger.write('Checking for updates...\n');

  const currentVersion = getCurrentVersion();
  const latestVersion = await getLatestVersion();

  const isCurrentVersionOutdated = semver.gt(latestVersion, currentVersion);

  if (isCurrentVersionOutdated) {
    console.log(warning(asterisks));
    console.log(warning`
      There is a new version available ({green.bold ${latestVersion}}).
      You can update the CLI version with the following command:

          ${code`npm i -g @corva/create-app@latest`}

    `);
    console.log(warning(asterisks));
  } else {
    logger.write('  ✅  \n');
  }
}

const PROMPT_MESSAGE = `Bumping package version:
${chalk.bold` Please select one of the options below:`}
  `;

export const ensureBumpVersion = async (bumpVersion) => {
  if (bumpVersion) {
    return bumpVersion;
  }

  const { option, custom } = await inquirer.prompt([
    {
      message: PROMPT_MESSAGE,
      name: 'option',
      type: 'list',
      choices: [
        { value: 'major', name: 'major (x+1.y.z)' },
        { value: 'minor', name: 'minor (x.y+1.z)' },
        { value: 'patch', name: 'patch (x.y.z+1)' },
        new inquirer.Separator(),
        { value: 'custom', name: 'Enter a custom version' },
        { value: 'skip', name: 'Skip version bump' },
      ],
      default: 'patch',
    },
    {
      message: 'Enter custom version',
      name: 'custom',
      when: (answers) => answers.option === 'custom',
      type: 'input',
      validate: (input) => (semver.valid(input) ? true : 'Invalid semver version'),
    },
  ]);

  return option === 'custom' ? custom : option;
};

export async function getIncreasedVersion(version, { bumpVersion: releaseTypeOrNewVersion }) {
  if (releaseTypeOrNewVersion === 'skip') {
    return version;
  }

  if (semver.valid(releaseTypeOrNewVersion)) {
    return releaseTypeOrNewVersion;
  }

  return semver.inc(version, releaseTypeOrNewVersion);
}
