// @corva/create-app/lib/flows/lib/manifest.js
import { APP_TYPES, APP_RUNTIMES } from '../../constants/cli.js';

const NODE_RUNTIMES = [APP_RUNTIMES.NODE12, APP_RUNTIMES.NODE14, APP_RUNTIMES.NODE16];

export class Manifest {
  constructor(manifest) {
    this.manifest = manifest;
  }

  get runtime() {
    return this.manifest.settings.runtime;
  }

  get type() {
    return this.manifest.application.type;
  }

  get name() {
    return this.manifest.application.name;
  }

  get unix_name() {
    // todo: npm scopes?
    return this.manifest.application.key.replace(/\W+/gi, '_').toLowerCase();
  }

  get description() {
    return this.manifest.application.description;
  }

  isNode() {
    return NODE_RUNTIMES.includes(this.manifest.settings.runtime);
  }

  isUi() {
    return this.manifest.application.type === APP_TYPES.UI;
  }

  isJs() {
    return this.isUi() || this.isNode();
  }

  isPython() {
    return this.manifest.settings.runtime.startsWith('python');
  }

  isScheduler() {
    return this.manifest.application.type === APP_TYPES.SCHEDULER;
  }
}
